apps/hannk/halide/add_generator.cpp

```diff
@@ -0,0 +1,61 @@
+#include "Halide.h"
+#include "common_halide.h"
+
+using namespace Halide;
+using namespace Halide::ConciseCasts;
+
+namespace hannk {
+
+class Add : public Generator<Add> {
+public:
+    // Input buffers and quantization parameters.
+    Input<Buffer<uint8_t>> input1_{"input1", 2};
+    Input<uint8_t> input1_zero_{"input1_zero"};
+    Input<int32_t> input1_multiplier_{"input1_multiplier"};
+    Input<uint32_t> input1_shift_{"input1_shift"};
+
+    Input<Buffer<uint8_t>> input2_{"input2", 2};
+    Input<uint8_t> input2_zero_{"input2_zero"};
+    Input<int32_t> input2_multiplier_{"input2_multiplier"};
+    Input<uint32_t> input2_shift_{"input2_shift"};
+
+    // Offset, quantization multiplier and shift for the output.
+    Input<uint8_t> output_zero_{"output_zero"};
+    Input<int32_t> output_multiplier_{"output_multiplier"};
+    Input<uint32_t> output_shift_{"output_shift"};
+    Input<uint8_t> output_min_{"output_min"};
+    Input<uint8_t> output_max_{"output_max"};
+
+    Output<Buffer<uint8_t>> output_{"output", 2};
+
+    void generate() {
+        Var x("x"), y("y");
+
+        Expr input1 = i32(i16(input1_(x, y)) - i16(input1_zero_)) << 20;
+        Expr input2 = i32(i16(input2_(x, y)) - i16(input2_zero_)) << 20;
+
+        input1 = rounding_shift_right(multiply_2x_high(input1, input1_multiplier_), input1_shift_);
+        input2 = rounding_shift_right(multiply_2x_high(input2, input2_multiplier_), input2_shift_);
+
+        Expr output = multiply_2x_high(input1 + input2, output_multiplier_);
+        output = i16_sat(rounding_shift_right(output, output_shift_));
+        output = u8_sat(saturating_add(output, output_zero_));
+        output_(x, y) = clamp(output, output_min_, output_max_);
```

Comment:
Looks like this sequence is repeated in many of the ops, maybe worth creating a helper function in common_halide?

Reply:
I am on the fence about this one. There actually was a helper function that did more of these steps, but I removed it. Some of the instances of it are slightly different, and there were a few cases where it should be different, and it's only 4 lines of code.